### PR TITLE
feat(workflows): include: load-time workflow inlining primitive (#2121)

### LINKED ISSUE #2121: feat(workflows): sub-workflow composition — include: (load-time inlining) and workflow: (runtime sub-run) primitives
## Problem — proven by audit

Composability currently stops at the command file: a workflow cannot reuse another workflow's DAG structure. The 2026-07-15 defaults audit quantified the consequence:

- The **9-node review block** (scope → sync → 5 review agents → synthesize → implement-fixes) is copy-pasted verbatim into 4 bundled workflows (idea-to-pr, plan-to-pr, issue-review-full, comprehensive-pr-review) and gated in a 5th (smart-pr-review)
- `verify-pr-base` is a byte-identical 15-line bash node in up to 9 workflows
- Clone workflows exist BECAUSE extraction is impossible: plan-to-pr is 17/18 nodes identical to idea-to-pr

Every copy drifts independently. This is the load-bearing composability gap.

### Phase 1: `include:` — load-time inlining (recommended first)

A node that references another workflow file; the loader expands its nodes into the parent DAG at parse time.

```yaml
- id: review
  include: archon-review-block        # resolved like command:/workflow discovery
  depends_on: [create-pr]
  with:                               # optional input mapping → variables inside the included DAG
    pr_ref: $create-pr.output.number
```

Semantics (reuse the loop_group body precedent deliberately):
- Included nodes are a **sealed sub-DAG** namespaced `<includeId>.<nodeId>` (id-shadowing check + event step_names already exist for loop_group bodies — #2090 shipped the namespacing plumbing)
- From the parent's perspective the include behaves like one node: parent `depends_on`/`when`/`trigger_rule` attach to it as a whole; `$includeId.output` = the included DAG's terminal-node output (same selection rule as loop_group)
- Load-time expansion means ZERO new runtime machinery: one run record, existing resume/skip/events/approval behavior all work unchanged
- Validation: cycle/uniqueness checks recurse (loader already does this for loop_group); include depth capped (e.g. 3); no recursive self-include
- Discovery/precedence identical to workflows today (bundled < global < project)

This alone deletes the review-block duplication: extract it once as `archon-review-block.yaml`, include it from 5 workflows.

### Phase 2: `workflow:` — runtime sub-run invocation (design with #1764)

A node that starts a **child workflow run**: own run record, own artifacts dir, own approval gates, parent consumes the child's terminal output. True 'run a workflow inside a workflow'.

Strictly heavier: nested lifecycle (cancel/pause propagation), resume across parent+child, path-lock interaction, dock/UI nesting, cost attribution. It overlaps the orchestrator-worker fan-out design (#1764) — a dynamic worker IS a sub-run — so phase 2 should be specified jointly with that issue rather than independently.

## Recommendation

Ship phase 1 (`include:`) — it solves ~90% of the observed duplication with loader-level work and the namespacing groundwork from #2090 already in place. Treat phase 2 as a #1764 co-design.

## References

- Audit: .claude/archon/reviews/2026-07-15-defaults-audit.md (session-local; findings summarized above)
- loop_group body sealing/namespacing precedent: packages/workflows/src/schemas/dag-node.ts, loader.ts validateDagStructure recursion, dag-executor.ts stepNamePrefix (#2090)
- Related: #1764 (dynamic fan-out), #2103 (workflow-builder schema)

### PR DESCRIPTION
**Pull request by Wirasm**
## Summary

- **Problem:** A workflow could not reuse another workflow's DAG structure. The 9-node PR review block (`verify-pr-base → review-scope → sync → 5 review agents → synthesize → implement-fixes`) was copy-pasted verbatim into multiple bundled workflows, and every copy drifted independently.
- **Why it matters:** Editing the shared review flow meant hand-editing several files and silently missing one. Composability stopped at the command file.
- **What changed:** Added an `include: <workflow-name>` DAG node mode that the loader expands at **discovery time**, inlining the referenced workflow's nodes into the parent DAG as a flattened, namespaced sub-DAG. Extracted `archon-review-block.yaml` and replaced the pasted block in 3 workflows with a single `include` node each.
- **What did NOT change (scope boundary):** No new runtime machinery — expansion completes before the executor ever sees the DAG, so one run record, and events/resume/skip/approvals all work unchanged on the namespaced nodes. **Phase 2** (`workflow:` runtime sub-runs — a child run with its own run record/artifacts/approvals) is explicitly **out of scope** per the issue; it will be co-designed with #1764.

### After

```
Author edits the review flow in ONE file (archon-review-block.yaml)
   │
   ▼
archon-idea-to-pr / archon-plan-to-pr / archon-issue-review-full
   each contain just:
     - id: review
       include: archon-review-block
       depends_on: [<upstream>]
   │
   ▼ AT DISCOVERY (load time)
   review  →  review__verify-pr-base → review__review-scope → review__sync
              → [5 review agents] → review__synthesize → review__implement-fixes
   │
   ▼ executor / events / resume see ordinary flat top-level nodes
   $review.output  ==  review__implement-fixes output (the sub-DAG terminal)
```

### After

```
 parseWorkflow (per-file)              discoverWorkflows
   dagNodeSchema [~ include mode]  ──▶   assemble name→workflow map
   validateDagStructure [~]                    │
   (rejects include-in-loop_group)             ▼
                                        [+] expandWorkflowIncludes  ← include-expander.ts
                                              recursion / depth cap / cycle detect
                                              namespacing + edge rewrite
                                              re-run validateDagStructure
                                                    │
                                                    ▼ (zero include nodes remain)
                                        executeDagWorkflow [~ fail-fast guard]
```

**Connection inventory:**

| From | To | Status | Notes |
|------|----|--------|-------|
| `workflow-discovery` | `include-expander` | **new** | final discovery step; both return branches |
| `include-expander` | `loader.validateDagStructure` | **new** | re-validate the flattened DAG (`validateDagStructure` now exported) |
| `loader.parseDagNode` | `schemas` (`isIncludeNode`, `INCLUDE_NODE_IGNORED_FIELDS`) | **new** | ignored-field warning for include nodes |
| `loader.validateDagStructure` | — | **modified** | rejects `include` inside a `loop_group` body |
| `dag-executor` | `schemas.isIncludeNode` | **new** | fail-fast guard (invariant: no include node reaches the executor) |
| `validator.validateWorkflowResources` | — | **modified** | skips include nodes in the raw `POST /validate` path |

## Label Snapshot

- Risk: `risk: low`
- Size: `size: L`
- Scope: `workflows`
- Module: `workflows:loader`, `workflows:include-expander`

## Change Metadata

- Change type: `feature`
- Primary scope: `workflows`

## Linked Issue

- Closes #2121
- Related #2090 (loop_group event namespacing — precedent), #2032 (loop_group node), #1764 (dynamic fan-out — Phase 2 co-design)

## Validation Evidence (required)

```bash
bun run validate
# → exit 0
#   check:bundled            up to date (36 commands, 21 workflows)
#   check:bundled-skill      up to date
#   check:bundled-schema     up to date
#   check:pi-vendor-map      OK
#   type-check / lint / format:check  clean
#   tests: 5706 passed, 0 failed across the monorepo
```

- Evidence provided: full `bun run validate` run (exit 0, zero test failures). New tests: 6 schema cases, 3 loader cases, 15 expander cases, 3 end-to-end executor cases.
- Real-discovery acceptance: `bun run cli validate workflows archon-review-block | archon-idea-to-pr | archon-plan-to-pr | archon-issue-review-full` → all `ok`. A discovery smoke test confirmed each includer expands to 10 `review__*` namespaced nodes, entry-node `depends_on` rewired to the upstream, and downstream `depends_on` rewired to the block sink `review__implement-fixes`.
- Intentionally skipped: a live `bun run cli workflow run archon-issue-review-full` (Level 4b) — it would execute real bash/gh/AI against a PR and burn cost; the discovery + validate evidence proves the namespaced structure instead.

## Security Impact (required)

- New permissions/capabilities? **No**
- New external network calls? **No**
- Secrets/tokens handling changed? **No**
- File system access scope changed? **No**

Load-time-only graph transformation on already-parsed workflow definitions. Include targets are literal names resolved through existing discovery precedence — no templated/dynamic/cross-repo targets.

## Compatibility / Migration

- Backward compatible? **Yes** — `include` is a new optional node mode; existing workflows are untouched. Include-free workflows pass through the expander byte-for-byte (fast path returns the raw object).
- Config/env changes? **No**
- Database migration needed? **No**

## Human Verification (required)

- Verified scenarios: all 4 CLI workflow validations pass through real discovery + expansion; discovery smoke test asserts exact namespaced ids, rewired entry/downstream edges, and terminal-output rewrite; end-to-end executor run asserts namespaced `step_name`s in the event log, `$inc.output` resolving to the child terminal, resume-skip of a namespaced node, and `always_run` re-execution surviving inlining.
- Edge cases checked (tests): unknown target, self-include cycle, mutual include cycle, depth-cap chain, namespaced-id collision with a hand-written node, two includes of the same block (distinct namespaces), `when`/`trigger_rule` propagation to entry nodes, nested includes, `with:` rejection, `include`-in-`loop_group` rejection, determinism (expand twice → identical), input-map non-mutation.
- What was not verified: a live end-to-end `workflow run` against a real GitHub PR (cost/side-effects); the web workflow builder round-trip (see follow-ups).

## Side Effects / Blast Radius (required)

- Affected subsystems/workflows: `@archon/workflows` discovery/loader/validator/executor; the 3 rewritten bundled workflows + the new `archon-review-block`; the embedded bundle.
- Potential unintended effects: any workflow with a top-level include now runs through the expander. Mitigated by the include-free fast path (untouched objects) and re-running `validateDagStructure` on the flattened result. Refs to include IDs inside **command files** (not node text) are not rewritten — an inherent limitation; none of the converted workflows rely on it.
- Guardrails: per-workflow error isolation (a bad include drops only that workflow); executor fail-fast if an unexpanded include node ever reaches it; `check:bundled` proves the regenerated bundle matches.

## Rollback Plan (required)

- Fast rollback: revert the 4 commits on `feat/include-primitive` (engine / defaults / tests / docs). The defaults commit is independently revertible to restore the pasted blocks while keeping the primitive.
- Feature flags: none — the feature is inert unless a workflow uses `include:`.
- Observable failure symptoms: a workflow with an include appearing in `/workflow list` errors ("include target not found" / "cycle" / "depth limit" / "Duplicate node id"), or the executor throwing "include node reached the executor unexpanded".

## Risks and Mitigations

- Risk: Dotted namespace would break `$id.output` substitution (the issue sketched `<includeId>.<nodeId>`).
  - Mitigation: **Deviation from the issue text** — the delimiter is `__` (double underscore), not `.`. The output-ref substitution regex forbids dots in a node id, so a dotted id would silently break every rewritten internal reference; `__` is inside the legal id class. Covered by a test asserting a rewritten internal ref resolves.
- Risk: Non-deterministic expansion would break resume (namespaced ids must match persisted `step_name`s byte-for-byte).
  - Mitigation: expansion is pure and ordered; determinism test + end-to-end resume-skip test. Every execution path re-discovers → re-expands.
- Risk: `archon-comprehensive-pr-review` looks like a 4th conversion candidate but is **not** byte-identical (no `verify-pr-base`, entry named `scope`).
  - Mitigation: **Deviation from the plan** — converted 3 workflows, not 4. Including the block there would silently add a PR-base retarget that hard-fails when no open PR exists (a behavior change, not dedup). Deferred with `archon-smart-pr-review`.

### Follow-ups (out of scope here)

- Regenerate `packages/web/src/lib/api.generated.d.ts` so the OpenAPI `DagNode` surfaces the new optional `include` field for the workflow builder. Root `bun run type-check` passes without it (no web code consumes `include`); regen needs a running server on :3090.
- Convert `archon-comprehensive-pr-review` / `archon-smart-pr-review` if/when their (differing/gated) review sub-DAGs are reconciled with the shared block.
- Phase 1b `with:` input mapping; Phase 2 `workflow:` runtime sub-runs (#1764).

## Review Fixes

Deep-review findings addressed in follow-up commits `6c30348e` and `fc2c0375` (validate stays green, 5727 tests):

Round 1 (`6c30348e`):
1. Moved the `isIncludeNode` executor guard to the first statement of the per-node body (was after resume-skip/`when:`/trigger, so an unexpanded include could slip through by matching a prior-completed entry or false gate). +2 tests.
2. Combine an include-level `when:` with an entry node's own `when:` as `A && B`, but only when neither side uses `||` (the grammar has no parentheses and `&&` binds tighter); otherwise fail at load time. +5 tests.
3. Error duplicate workflow `name:`s across different files (was silent last-wins in the include name-map). +1 test.

Round 2 (`fc2c0375`):
4. **Shorthand `when:` refs** (`$id.field` == `$id.output.field`) are now rewritten during namespacing — previously only literal `.output` refs were, so `when: "$verify.exit_code == '0'"` silently pointed at a renamed sibling (fail-closed skip). Both directions covered.
5. **Block command files scanned at load time.** A `command:` node's `.md` is read at execution time and can't be rewritten, so a `$sibling.output` inside it would silently resolve to `''`. Discovery now pre-resolves include-target command content and the expander scans it (fences stripped): a sibling ref FAILS expansion; an unresolvable file WARNs. (The extracted review block references no siblings — verified.)
6. **`validateDagStructure` now scans every substituted surface** — bash, script, `approval.message`, cancel, `loop.until_bash`, `loop_group.until_bash` — not just when/prompt/loop.prompt. Widens validation for all workflows; the full bundled set + suite pass. Two repo fixtures had a placeholder `$node.output` in a code comment — fixed the comments, not the check.
7. **Fence-aware prose rewriting** — a `$sibling.output` inside a fenced example in a prompt/`loop.prompt`/`approval.message` is left verbatim while live refs are renamed; code fields stay blind.
8. **Dropped workflow-level fields now WARN** at load time (requires/persist_sessions/interactive/worktree/provider/model), calling out `requires:['github']`. Necessary sub-fix: the loader was **not parsing `requires` at all** — added it (warn-and-drop, mirroring `tags`), which both enables this warn and fixes a latent gap where the runtime github capability gate could never fire for a discovered workflow.

## Summary by CodeRabbit

* **New Features**
  * Added load-time `include:` nodes to inline reusable workflow blocks as namespaced sub-DAGs with `$<includeId>.output` terminal resolution.
  * Added optional `requires` support during workflow parsing.
  * Refactored default PR review workflows to use a shared review orchestration block.
* **Documentation**
  * Documented `include:` expansion, namespacing/rewiring, and limitations (including updated schema quick reference).
* **Bug Fixes**
  * Fail-fast when an unexpanded `include` reaches execution.
  * Improved load-time reference validation and skip resource validation for include nodes; reject `include` inside `loop_group`.
* **Tests**
  * Expanded include-expansion, resume/skip, determinism, loader validation, and executor behavior coverage.

### CODE REVIEW

File: packages/workflows/src/dag-executor.ts [hunk 1]
```diff
@@ -4434,6 +4435,16 @@ async function runLayers(ctx: RunLayersContext): Promise<void> {
...
 
+          if (isIncludeNode(node)) {
+            // Include nodes are expanded away at discovery time (include-expander.ts):
+            // one must never reach the executor. If it does, discovery was bypassed —
+            // fail loud rather than silently mis-running it as a prompt node.
+            throw new Error(
+              `Internal error: include node '${node.id}' reached the executor unexpanded. ` +
+                'Include nodes must be resolved by expandWorkflowIncludes() during discovery.'
+            );
+          }
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply packages/workflows/src/dag-executor.ts]: Fixed in 6c30348e — moved the `isIncludeNode` guard to the very first statement of the per-node body, before resume-skip / `when:` / trigger-rule handling. Added two tests (`executeDagWorkflow -- unexpanded include node fail-fast guard`): an unexpanded include node with a `priorCompletedNodes` entry, and one with a false `when:`, both now fail loudly (`node_failed`, error "reached the executor unexpanded") instead of silently skipping.

File: packages/workflows/src/include-expander.ts [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    rewriteNodeOutputRefs(clone, rename);
+    clone.id = prefix + cn.id;
+
+    if (wasEntry) {
+      // Entry node: the include node's upstream deps + gate attach here. The gate
+      // (when/trigger_rule) copies onto each entry unless the entry declares its own.
+      if (parentDeps.length > 0) clone.depends_on = [...parentDeps];
+      if (includeNode.when !== undefined && clone.when === undefined) {
+        clone.when = includeNode.when;
+      }
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply packages/workflows/src/include-expander.ts]: Fixed in 6c30348e, but with a different combiner than proposed. The `when:` grammar has **no parentheses** — condition-evaluator.ts splits on `||` first, then `&&`, with no grouping. So the suggested `(${includeNode.when}) && (${clone.when})` would be unparseable and fail-closed to a silent skip (a behavior change).

Instead I combine as plain `${includeNode.when} && ${clone.when}`, which is only semantics-preserving when **neither** side contains `||` (since `&&` binds tighter than `||`, plain concatenation preserves `(A) && (B)` only without `||`). If either side uses `||`, the expansion now **fails at load time** with a clear message telling the author to restructure (move the gate onto the include node, or gate inside the block) — fail-fast beats silently-wrong precedence.

Tests added: both-plain combine, entry-side `||` → load error, include-side `||` → load error, entry-only unchanged, include-only unchanged.

File: packages/workflows/src/workflow-discovery.ts [hunk 3]
```diff
@@ -210,6 +211,31 @@ export async function discoverWorkflows(
...
+    const { workflows: expandedByName, errors: expansionErrors } =
+      expandWorkflowIncludes(rawByName);
+    allErrors.push(...expansionErrors);
+
+    const result: WorkflowWithSource[] = [];
+    for (const { workflow, source } of workflowsByFile.values()) {
+      const expanded = expandedByName.get(workflow.name);
+      if (!expanded) continue; // expansion failed for this workflow — drop it
+      result.push({ workflow: expanded, source });
+    }
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply packages/workflows/src/workflow-discovery.ts]: Fixed in 6c30348e — discovery now detects duplicate workflow `name:`s among surviving (post-filename-precedence) files and errors the colliding entries with a `WorkflowLoadError` (resilient: only the collisions are dropped and errored, the rest still load). The colliding names are also excluded from the include name-map so resolution can no longer silently pick a last-wins winner.

Worth noting this hardens a **pre-existing** ambiguity: same-name-in-different-files was already order-dependent before `include:` (two `WorkflowWithSource` entries with the same name) — the include name-map just made it load-bearing. Verified no false positives on the bundled set (0 duplicate-name errors across the 34 discovered workflows). Test added: two files, same `name:`, different filenames → both errored, neither surfaces as usable.

**Comment by Wirasm** [comment]
## Deep-review fixes (commit `fc2c0375`)

Addressed all 5 findings from the deep review. `bun run validate` stays green (exit 0, 5727 tests, +13 new; all generated-file checks up to date).

**CRITICAL 1 — shorthand `when:` refs rewritten.** The condition grammar accepts `$id.field` as shorthand for `$id.output.field`; the rewriter (keyed on literal `.output`) missed it, so an included node's `when: "$verify.exit_code == '0'"` kept pointing at the pre-rename sibling → silent fail-closed skip. Added a `when:`-specific rename covering both the internal-sibling and downstream `$includeId.field` directions.

**CRITICAL 2 — block command files scanned at load time.** A `command:` node's `.md` is read at execution time and can't be rewritten, so a `$sibling.output` inside it would silently substitute to `''`. Discovery now pre-resolves include-target command content (repo → `~/.archon/commands` → bundled) and the expander scans it (fences stripped): a sibling ref FAILS expansion; an unresolvable file WARNs. The extracted review block's command files reference no siblings (verified), so this is a forward guard.

**HIGH — `validateDagStructure` scans every substituted surface.** It only covered when/prompt/loop.prompt, but the executor also substitutes bash, script, `approval.message`, cancel, `loop.until_bash`, `loop_group.until_bash`. Extended the scan to all of them (prose keeps fence-stripping; code fields verbatim). This widens validation for ALL workflows — the full bundled set + suite pass. Two repo fixtures had a placeholder `$node.output` in a bash/script comment; I fixed the comments (→ `$<node>`), not the check.

**MEDIUM 1 — fence-aware prose rewriting.** A `$sibling.output` inside a fenced/inline-code example in prompt/loop.prompt/approval.message is now left verbatim while live refs in the same field are renamed (matching the loader's fence-stripping). Code fields stay blind — their refs are live.

**MEDIUM 2 — dropped workflow-level fields WARN.** Inlining drops the block's requires/persist_sessions/interactive/worktree/provider/model; the expander now emits a one-line load-time warn listing them, calling out `requires:['github']`. Necessary sub-fix I found while verifying: the loader was **not parsing `requires` at all** (silently dropped at parse), so both this warn and the runtime github capability gate were blind to it — added `requires` parsing (warn-and-drop, mirroring `tags`).

Note on the finding-2 combiner: I used plain `A && B` (grammar has no parentheses; `&&` binds tighter than `||`), failing fast when either side uses `||` — the parenthesized suggestion would be unparseable.

**Comment by Wirasm** [comment]
Maintainer decision: the `requires:` loader-parsing fix stays in this PR. Rationale: it's a correctness fix toward already-documented behavior (`requires: [github]` has been in CLAUDE.md and the workflow schema all along — the loader just never carried the field through, so the gate was inert for every discovered workflow), it's a no-op on solo installs (gate activates only under App mode + TOKEN_ENCRYPTION_KEY), and the include-expander's dropped-fields warning depends on the field existing. Flagged as a runtime behavior change in the PR body — reviewers of App-mode deployments should note that workflows declaring `requires: [github]` will now actually block unconnected users, as designed.

File: packages/workflows/src/include-expander.ts [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+      droppedFields,
+      ...(child.requires?.includes('github')
+        ? {
+            note: "requires:['github'] is dropped by inlining — declare it on the PARENT workflow if the block needs GitHub identity",
+          }
+        : {}),
+    },
+    'include.workflow_level_fields_dropped'
+  );
+}
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply packages/workflows/src/include-expander.ts:300]: Fixed in 51953a47. Rewrote `warnDroppedWorkflowLevelFields` to **derive** the dropped set from the child `WorkflowDefinition`'s own defined keys rather than a hand-list (repo convention: no duplicates of schema-derived fields), so `mutates_checkout`, `effort`, `thinking`, `modelReasoningEffort`, `webSearchMode`, `fallbackModel`, `betas`, and `sandbox` are now covered automatically — and any future workflow-level field will be too.

- Excluded `name`/`description`/`nodes` (identity / the thing being inlined) and `tags` (cosmetic keyword-inference metadata, behaviorally inert — commented why).
- Filtered out the always-present-but-`undefined` keys `parseWorkflow` emits (provider/model/modelReasoningEffort/webSearchMode/interactive), so an unset field is not reported as dropped.
- Kept the explicit `requires:[github]` callout and added an equally explicit `safetyNote` when `mutates_checkout` or `sandbox` are dropped (the safety/isolation-relevant ones).

Tests: updated the existing M2 test (asserts undefined keys are filtered and no safety note when none apply) and added one asserting a block with `mutates_checkout: false` + `sandbox` warns and names both in the safety note. `bun run validate` green (5728 tests).

**Comment by Wirasm** [comment]
## PR Review Summary

Multi-agent review (7 specialists: code-reviewer, silent-failure-hunter, pr-test-analyzer, type-design-analyzer, comment-analyzer, docs-impact, code-simplifier) at head `51953a47`.

> Note: commit `51953a47` landed mid-review and already resolved one finding — the dropped-workflow-level-field warning previously missed `mutates_checkout`/`sandbox` (a silent safety-flag loss on include); it now derives the warn set generically from the child's keys, with tests. Validation was re-run at the new head.

### Critical Issues (0 found)

None.

### Important Issues (5 found)

| ID | Agent | Issue | Location |
|----|-------|-------|----------|
| I1 | code-reviewer | Depth cap contradicts docs: `stack.length >= INCLUDE_MAX_DEPTH` (3) errors on the 3rd nesting hop, but the guide promises "includes may nest up to 3 levels deep" — verified empirically (A→B→C passes, A→B→C→D fails). The 4-hop test pins the off-by-one, suggesting it wasn't noticed. Fix the doc (lower risk) or loosen to `>` | `include-expander.ts:55,338` vs `authoring-workflows.md:849` |
| I2 | silent-failure-hunter + pr-test-analyzer | `$LOOP_PREV.<id>.output` refs are not rewritten during include namespacing (neither ref regex matches that grammar), so a loop_group body inside an included block referencing a renamed sibling silently resolves to `''` every iteration (debug-level log only, indistinguishable from first-iteration). Not exploitable by the shipped review block, but loop_group-inside-include has zero test coverage. Rewrite the ref or fail-loud reject the pattern | `include-expander.ts:64,74,135`; runtime swallow at `dag-executor.ts:633` |
| I3 | comment-analyzer + code-simplifier | `resolveCommandContentForScan` claims to mirror the validator search order but calls `getCommandFolderSearchPaths()` without `config.commandFolder` and ignores `defaults.loadDefaultCommands: false` — for repos with a custom `commands.folder`, the sibling-ref safety scan silently degrades to a WARN-and-skip exactly where it's needed. It also re-implements ~40 lines of `loadCommandPrompt`'s search walk; extract a shared dir-walk helper | `workflow-discovery.ts:190-238` vs `executor-shared.ts:247-384` |
| I4 | pr-test-analyzer | Approval gates inside an included block are untested (criticality 7/10): nothing proves gate resolution (`/workflow approve`) or `capture_response` output threading survives namespacing, especially with the same approval-bearing block included twice (cross-talk risk) | `include-expander.test.ts`, `dag-executor.test.ts` |
| I5 | comment-analyzer + silent-failure-hunter | validator comment is factually wrong: `POST /api/workflows/validate` never calls `validateWorkflowResources` (it only calls `parseWorkflow`), so the `isIncludeNode` skip is currently unreachable defensive code, not the "raw validate path" the comment claims. Reword as defensive-only or wire the endpoint | `validator.ts:388-393` |

### Suggestions (7 found)

| ID | Agent | Suggestion | Location |
|----|-------|------------|----------|
| S1 | type-design-analyzer | Optional hardening: a `PostExpansionWorkflow` type (`nodes: Exclude<DagNode, IncludeNode>[]`) as the expander's return type and the executor entry-point's input — compile-time regression guard at the one real boundary, without threading through loop_group internals. Ratings: encapsulation 6/10, invariant expression 3/10, usefulness 8/10, enforcement 6/10 — runtime-only guard judged pragmatic and consistent with house style | `include-expander.ts`, `dag-executor.ts` |
| S2 | code-simplifier | The include branch of the `dagNodeSchema` transform re-derives 5 structural fields byte-for-byte instead of reusing a shared `structural` subset of `base` | `dag-node.ts:713-724` vs `791-800` |
| S3 | code-simplifier | Use the existing `isBashNode()` guard instead of `'bash' in node && typeof node.bash === 'string'` — every adjacent branch already uses guards | `loader.ts:236` |
| S4 | code-simplifier | Rule of Three now met: three independent enumerations of "which fields carry `$id.output` refs" (loader validation, expander rewrite, executor substitution call sites). Extract a shared field descriptor or add explicit keep-in-sync comments at all three sites | `loader.ts:230-246`, `include-expander.ts:135-161`, `dag-executor.ts` |
| S5 | pr-test-analyzer | Add composition tests: `persist_session` + include (two copies → independent session rows) and diamond includes (two blocks including the same leaf), both criticality 5/10; also the cycle-check-before-memo ordering invariant beyond the linear case | `include-expander.test.ts` |
| S6 | comment-analyzer | Stale/inconsistent docblocks: `isPersistableNode` docstring omits the new include exclusion; three docblocks give three different "structural fields" lists for include nodes, and the transform comment wrongly claims the expander reads `description` | `dag-node.ts:861-868`, `487-493`, `785-790` |
| S7 | silent-failure-hunter | Expansion errors set `filename: <workflow name>` where every other `WorkflowLoadError` carries a real path — field-semantics inconsistency for future tooling | `include-expander.ts:445` |

### Strengths

- Executor fail-fast guard is the **first** statement in the per-node body — before resume-skip/`when:`/trigger-rule — with two adversarial tests proving an unexpanded include can't slip through a stale resume entry or false gate.
- Ref rewriting handles the hard subtleties: fence-aware prose (examples left verbatim), `when:` shorthand `$id.field` grammar, and the `__` delimiter choice is well-justified (dotted ids would silently break every substitution regex).
- Determinism test (byte-identical re-expansion) is load-bearing for resume correctness; the "every path re-discovers → re-expands" claim was independently traced across CLI, API, and chat resume paths and holds.
- Per-workflow error isolation: a bad include (unknown target / cycle / depth / id collision / broken command cross-ref) drops only that workflow with a loud, actionable error.
- Incidental latent-bug fix: `requires:` was never parsed by the loader since it was introduced (#1823), so the GitHub capability gate was silently inert for every discovered workflow — this PR fixes it.
- Honest scope boundaries: `with:` and include-in-loop_group are rejected loudly, and the command-file ref limitation is documented in the PR body rather than left to be discovered.

### Documentation Issues

- `CLAUDE.md:818` — the node-type enumeration (`command:`/`prompt:`/`bash:`/`loop:`/…) was historically updated for every new node type; `include:` is missing.
- `packages/docs-web/src/content/docs/book/quick-reference.md` (~line 120) — the "Node Options (DAG)" one-of table lacks an `include` row.
- `authoring-workflows.md:826-834` — the new section doesn't disclose the sink asymmetry: `depends_on: [includeId]` fans out to **all** sinks while `$includeId.output` resolves only to the **primary (first)** sink; add a sentence for multi-terminal blocks.
- `authoring-workflows.md:849` — the "3 levels" nesting claim (see I1).
- The rest of the new include docs were verified line-by-line against the implementation and are accurate; none of the pages describing the 3 rewritten bundled workflows go stale (they describe behavior, not node structure).

### Validation Results

| Check | Result | Notes |
|-------|--------|-------|
| Type check | PASS | all packages, at head `51953a47` |
| Tests (@archon/workflows) | PASS | all batches, at head `51953a47` |
| Lint | PASS | zero warnings |

### Verdict

**NEEDS FIXES** — nothing blocking, no critical issues. The core expansion engine is solid and unusually well self-reviewed (two prior fix rounds visible in the history). The remaining work is: reconcile the depth-cap doc/code mismatch (I1), decide the `$LOOP_PREV`/loop_group-in-include story (I2, rewrite or reject), thread `commandFolder` config into the command scan (I3), add the approval-in-include test (I4), and a batch of comment/doc corrections.

### Recommended Actions

1. Fix I1 (doc or check — doc fix is lower-risk) and I3 (real degraded-safety path for custom command folders).
2. Decide I2: add a `$LOOP_PREV` rewrite rule or fail-loud reject loop_group-bearing include targets that reference outer siblings.
3. Add the I4 approval-gate composition test before this pattern is used in a gated block.
4. Sweep the comment/doc items (I5, S6, Documentation Issues) in one commit.
5. Suggestions S1-S5/S7 are optional polish — reasonable to defer.

**Comment by Wirasm** [comment]
## Final-review round — disposition (commit `e0c410c8`)

`bun run validate` green (exit 0, 5735 tests, all generated-file checks up to date).

| ID | Verdict | Disposition |
|----|---------|-------------|
| **I1** depth cap vs docs | Valid | Fixed: loosened `>= INCLUDE_MAX_DEPTH` → `>` so exactly 3 levels are allowed (honors the documented "up to 3 levels"). Depth test updated to pin the boundary: a 3-level chain expands, 4-level fails. |
| **I2** `$LOOP_PREV` in included loop_group bodies | **Misdirected — non-bug** (evidence below) | No code change. Added the missing coverage: body `$LOOP_PREV.<id>` preserved + outer-sibling ref renamed; body-id shadow-collision rejected. |
| **I3** command-scan config parity | Valid | Fixed: thread `config.commands.folder` + `config.defaults.loadDefaultCommands` from `discoverWorkflowsWithConfig` into the scan. Custom-folder commands are now scanned (not degraded to WARN); bundled fallback skipped when defaults opted out. Test added. |
| **I4** approval-in-include tests | Valid (coverage gap) | Added: pause fires under the namespaced approval id; `capture_response` reachable via the namespaced id; same block twice → distinct `ApprovalContext.nodeId`, no cross-talk. |
| **I5** validator comment | Valid | Reworded as defensive-only (`POST /api/workflows/validate` only runs `parseWorkflow`, never the resource pass). Endpoint NOT wired (out of scope). |
| **S2** shared structural-field subset | Valid | Extracted `structuralBase` in dag-node's transform (was two byte-identical 5-field derivations). |
| **S3** use `isBashNode()` | Valid | Done at the loader ref-scan. |
| **S4** keep-in-sync comments | Valid (partial) | Added cross-referencing comments at all three ref-surface enumerations. Full shared-descriptor extraction **deferred** as a follow-up refactor (per your instruction). |
| **S5** composition tests | Valid | Added: `persist_session` survives inlining + independent per-parent keys; diamond include (two blocks → same leaf) expands without collision. |
| **S6** stale docblocks | Valid | Fixed: `isPersistableNode` now names the include exclusion; include transform no longer claims the expander "reads" description. |
| **S7** expansion-error filename | Valid | Fixed (short plumbing): errors re-keyed from workflow name → the includer's real filename in discovery. Tests updated. |
| **DOCS** | — | `include:` added to CLAUDE.md node-type list + quick-reference one-of table; sink-asymmetry documented in authoring-workflows (`depends_on` fans to ALL sinks; `$includeId.output` = primary/first sink). I1 doc already said "up to 3 levels" — code now matches. |
| **S1** compile-time `PostExpansionWorkflow` type | — | **Skipped** (per review): the runtime fail-fast guard + two adversarial tests already cover the boundary; threading a new type through the executor entry is disproportionate here. |

### I2 verdict (with evidence)

The reviewer's silent-`''` scenario for `$LOOP_PREV` in an included loop_group body is **unreachable**. Empirically, inlining a block whose loop_group body references both `$LOOP_PREV.inner.output` and an outer sibling `$seed.output`:

```
top-level ids: rev__seed, rev__lg
body node id:  inner            # NOT renamed (body-local)
body prompt:   "prev=$LOOP_PREV.inner.output outer=$rev__seed.output"
```

Body node ids are **not** renamed during inlining, so `$LOOP_PREV.<bodyId>` still points at the body node (correct, no rewrite needed). The outer-sibling ref (`$seed.output`) **is** rewritten to `$rev__seed.output` (already handled by the recursive rewrite). The real risk is a body id that collides with a parent top-level id post-inline — and the loader's enclosing-scope shadow check rejects it loudly (now covered by a test). So: no `$LOOP_PREV` rewrite is needed; the gap was **coverage**, not correctness.
